ethereum/contracts/SwapFactory.sol

```diff
@@ -59,25 +63,37 @@ contract SwapFactory is Secp256k1 {
         bytes32 _pubKeyRefund, 
         address payable _claimer, 
         uint256 _timeoutDuration,
+        address _asset,
+        uint256 _value,
         uint256 _nonce
     ) public payable returns (bytes32) {
 
         Swap memory swap;
-        swap.owner = payable(msg.sender); 
-        swap.claimer = _claimer;
+        swap.owner = payable(msg.sender);
         swap.pubKeyClaim = _pubKeyClaim;
         swap.pubKeyRefund = _pubKeyRefund;
+        swap.claimer = _claimer;
         swap.timeout_0 = block.timestamp + _timeoutDuration;
         swap.timeout_1 = block.timestamp + (_timeoutDuration * 2);
-        swap.value = msg.value;
+        swap.asset = _asset;
+        swap.value = _value;
+        if (swap.asset == address(0)) {
+            require(swap.value == msg.value, "value not same as ETH amount sent");
```

Comment:
alternatively, we can allow `_value` to be 0 for `swap.asset == address(0)` so we can remove this check and ignore `_value` for eth